python/semantic_kernel/utils/tracing/decorators.py

```diff
@@ -0,0 +1,233 @@
+                    _set_completion_response(
+                        span, completions, response_id or "unknown", prompt_tokens, completion_tokens
+                    )
+
+            return completions
+
+        return wrapper_decorator
+
+    return inner_trace_chat_completion
+
+
+def _start_completion_activity(
+    model_name: str, model_provider: str, chat_history: ChatHistory, execution_settings: PromptExecutionSettings | None
+) -> Span | None:
+    """Start a text or chat completion activity for a given model."""
+    if not are_model_diagnostics_enabled():
+        return None
+
+    operation_name: str = "chat.completions"
+
+    span = tracer.start_span(f"{operation_name} {model_name}")
+
+    # Set attributes on the span
+    span.set_attributes(
+        {
+            OPERATION: operation_name,
+            SYSTEM: model_provider,
+            MODEL: model_name,
+        }
+    )
+
+    if execution_settings:
+        attribute = execution_settings.extension_data.get("max_tokens")
+        if attribute:
+            span.set_attribute(MAX_TOKEN, attribute)
+
+        attribute = execution_settings.extension_data.get("temperature")
+        if attribute:
+            span.set_attribute(TEMPERATURE, attribute)
+
+        attribute = execution_settings.extension_data.get("top_p")
+        if attribute:
+            span.set_attribute(TOP_P, attribute)
+
+    if are_sensitive_events_enabled():
+        formatted_messages = _messages_to_openai_format(chat_history.messages)
+        span.add_event(PROMPT_EVENT, {PROMPT_EVENT_PROMPT: formatted_messages})
+
+    return span
+
+
+def _set_completion_response(
+    span: Span,
+    completions: list[ChatMessageContent],
+    response_id: str,
+    prompt_tokens: int | None = None,
+    completion_tokens: int | None = None,
+) -> None:
+    """Set the a text or chat completion response for a given activity."""
+    if not are_model_diagnostics_enabled():
+        return
+
+    span.set_attribute(RESPONSE_ID, response_id)
+
+    finish_reasons: list[str] = [str(content.finish_reason) for content in completions]
+    span.set_attribute(FINISH_REASON, ",".join(finish_reasons))
+
+    if are_sensitive_events_enabled() and completions:
+        span.add_event(COMPLETION_EVENT, {COMPLETION_EVENT_COMPLETION: _messages_to_openai_format(completions)})
+
+    if prompt_tokens:
+        span.set_attribute(PROMPT_TOKEN, prompt_tokens)
+
+    if completion_tokens:
+        span.set_attribute(COMPLETION_TOKEN, completion_tokens)
+
+
+def _set_completion_error(span: Span, error: Exception) -> None:
+    """Set an error for a text or chat completion ."""
+    if not are_model_diagnostics_enabled():
+        return
+
+    span.set_attribute(ERROR_TYPE, str(type(error)))
+
+    span.set_status(StatusCode.ERROR, str(error))
+
+
+def _messages_to_openai_format(messages: list[ChatMessageContent]) -> str:
```

Comment:
this function really needs to be redone, ChatMessageContents are pydantic models, and the pydantic json seralizer is faster then json.dumps, so at least switch to that, but more generally we can probably use one of the existing methods on chat history for this!

Reply:
Using ChatMessageContent.to_dict() now